distro/run/core/src/main/java/org/camunda/bpm/run/CamundaBpmRunConfiguration.java

```diff
@@ -16,64 +16,55 @@
 import org.springframework.context.annotation.Bean;
 import org.springframework.context.annotation.Configuration;
 
+import java.util.List;
+
 @EnableConfigurationProperties(CamundaBpmRunProperties.class)
 @Configuration
 @AutoConfigureAfter({ CamundaBpmAutoConfiguration.class })
 public class CamundaBpmRunConfiguration {
 
-  @Autowired
-  CamundaBpmRunProperties camundaBpmRunProperties;
-
   @Bean
   @ConditionalOnProperty(name = "enabled", havingValue = "true", prefix = CamundaBpmRunLdapProperties.PREFIX)
-  public LdapIdentityProviderPlugin ldapIdentityProviderPlugin() {
-    return camundaBpmRunProperties.getLdap();
+  public LdapIdentityProviderPlugin ldapIdentityProviderPlugin(CamundaBpmRunProperties properties) {
+    return properties.getLdap();
   }
 
   @Bean
   @ConditionalOnProperty(name = "enabled", havingValue = "true", prefix = CamundaBpmRunAdministratorAuthorizationProperties.PREFIX)
-  public AdministratorAuthorizationPlugin administratorAuthorizationPlugin() {
-    return camundaBpmRunProperties.getAdminAuth();
+  public AdministratorAuthorizationPlugin administratorAuthorizationPlugin(CamundaBpmRunProperties properties) {
+    return properties.getAdminAuth();
   }
 
   @Bean
-  public ProcessEngineConfigurationImpl processEngineConfigurationImpl(List<ProcessEnginePlugin> processEnginePlugins) {
-    final SpringProcessEngineConfiguration configuration = new CamundaBpmRunProcessEngineConfiguration();
-
-    // register process engine plugins defined in yaml
-    List<CamundaBpmRunProcessEnginePluginProperty> yamlPluginsInfo = camundaBpmRunProperties.getProcessEnginePlugins();
-    CamundaBpmRunProcessEnginePluginHelper.registerYamlPlugins(processEnginePlugins, yamlPluginsInfo);
+  public ProcessEngineConfigurationImpl processEngineConfigurationImpl(List<ProcessEnginePlugin> processEnginePluginsFromContext,
+                                                                       CamundaBpmRunProperties properties,
+                                                                       CamundaBpmRunDeploymentConfiguration deploymentConfig) {
+    String normalizedDeploymentDir = deploymentConfig.getNormalizedDeploymentDir();
+    boolean deployChangedOnly = properties.getDeployment().isDeployChangedOnly();
+    var processEnginePluginsFromYaml = properties.getProcessEnginePlugins();
 
-    configuration.getProcessEnginePlugins().add(new CompositeProcessEnginePlugin(processEnginePlugins));
-    return configuration;
+    return new CamundaBpmRunProcessEngineConfiguration(normalizedDeploymentDir, deployChangedOnly,
+        processEnginePluginsFromContext, processEnginePluginsFromYaml);
   }
 
   @Bean
-  public static CamundaDeploymentConfiguration camundaDeploymentConfiguration() {
-    return new CamundaBpmRunDeploymentConfiguration();
+  public CamundaBpmRunDeploymentConfiguration camundaDeploymentConfiguration(@Value("${camunda.deploymentDir:#{null}}") String deploymentDir) {
```

Comment:
❓ If we favour [Type-Safe Config Properties](https://docs.spring.io/spring-boot/reference/features/external-config.html#features.external-config.typesafe-configuration-properties) both for encapsulating injection & defaulting, should we create a separate Properties class for this property as well?

Reply:
❓ Is this also returning the env variable as before? `env.getProperty(CAMUNDA_DEPLOYMENT_DIR_PROPERTY)`

Reply:
Yes, below you can find a debugging with a breakpoint on the resolution of the property `camunda.deploymentDir`:

- On `master`, `Environment#getProperty`  resolves the property to the `resources` directory (see [screenshot](https://github.com/user-attachments/assets/ba9103c7-d484-4e46-9cce-adff16901f97))
- On `2652-add-deploy-changed-only-config`, injecting the property with `@Value` also returns the `resources` directory (see [screenshot](https://github.com/user-attachments/assets/bf150195-0677-43ae-bfaa-0bce5b44a056))

Reply:
As discussed on the [first point](https://github.com/camunda/camunda-bpm-platform/pull/4607#discussion_r1766578936), no separate type safe object was deemed necessary for `camunda.deploymentDir` property. The property can be handled with `@Value` for now and can be placed into a type safe object if more properties follow under the camunda namespace.